File: mamevert.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

#
# This requires nokogiri to parse the XML file so you
# might need to do:
#   sudo apt install ruby-nokogiri
#


require 'nokogiri'

# A class to parse MAME xml
class MameFilter < Nokogiri::XML::SAX::Document
  def start_document
    @units = []
    @items = []
  end

  def start_element(name, attrs = [])
    case name
    when 'machine'
      attrs = attrs.to_h
      @machine = attrs['name']

    when 'display'
      attrs = attrs.to_h
      if attrs['tag'].start_with?('screen') &&
         %w[270 90].include?(attrs['rotate'])
        puts @machine
      end
    end
  end
end

parser = Nokogiri::XML::SAX::Parser.new(MameFilter.new)
input = ARGV[0].nil? ? $< : File.new(ARGV[0],"r")
parser.parse(input)

File: mamevert.py
#!/usr/bin/env python3

import sys
import xml.sax

class MameFilter(xml.sax.ContentHandler):
    def __init__(self):
        self.current_machine = None

    def startElement(self, name, attrs):
        if name == 'machine':
            self.current_machine = attrs['name']
        elif name == 'display' and self.current_machine:
            if attrs.get('tag', '').startswith('screen') and attrs.get('rotate') in ['90', '270']:
                print(self.current_machine)

    def endElement(self, name):
        if name == 'machine':
            self.current_machine = None

if __name__ == "__main__":
    parser = xml.sax.make_parser()
    handler = MameFilter()
    parser.setContentHandler(handler)
    input_stream = open(sys.argv[1], 'r') if len(sys.argv) > 1 else sys.stdin
    parser.parse(input_stream)
